doc/DesignConsiderations.md

```diff
@@ -0,0 +1,120 @@
+
+  * Runtime headers (the `.crt0_header` section)
+  * Executable code (the `.text` section)
+  * Read-only data (the `.rodata` section)
+  * Non-zero-initialized read-write data (the `.data` section)
+
+A process' memory section is in RAM and contains:
+
+  * Heap memory (if a dynamic memory allocator is present)
+  * Zero-initialized read-write data (the `.bss` section)
+  * Non-zero-initialized read-write data (the `.data` section)
+  * Stack memory
+
+`libtock_core` developers should consider the space usage of their code, and
+where the space usage is (non-volatile memory or RAM). Important note: the
+`.data` section consumes space in both non-volatile memory and RAM!
+
+Most apps will not use all functionality in `libtock_core`. The size impact of
+using `libtock_core` in an app should be commensurate with the `libtock_core`
+functionality that app uses. If done correctly, this will allow `libtock_core`'s
+users to build Tock systems with multiple small, single-purpose apps without
+excessive code duplication.
+
+## Testability
+
+Programmatic tests are important to verify that new functionality works
+correctly and to keep existing functionality working through nontrivial
+refactorings. Different types of tests offer different capabilities:
+
+  * **Host-based unit tests** run on a non-Tock host OS (such as Linux). They
+    can only be used with portable `libtock_core` code, as they code they test
+    must run on both Tock and the host OS. These tests must either test code
+    that does not directly use system calls, or must direct the system calls to
+    a fake Tock kernel. These tests can be fast (compile and run in seconds),
+    and can easily simulate a variety of failure modes (e.g. kernel errors) that
+    may be difficult to generate with a real Tock kernel.
+  * **Emulation tests** run on an emulated Tock system. Currently, emulation
+    tests run in QEMU on an emulated HiFive1 Rev B board. These test
+    `libtock_core`'s interaction with a real Tock kernel, albeit with limited
+    hardware access.
+  * **Hardware tests** run on physical hardware with a real Tock kernel. These
+    are end-to-end tests that test not only `libtock_core`'s functionality but
+    also the Tock kernel's interaction with the real hardware. However,
+    automating these tests is difficult, and currently we do not have a way to
+    run hardware tests in CI.
+
+If `libtock_core` is functional on real hardware with a real kernel, it should
+be able to support emulation tests and hardware tests. However, host-based unit
+tests require more work. In its intended use case, `libtock_core` is not
+supposed to be portable, supporting only the Tock kernel. Hoever, host-based
+unit tests are very valuable, both for rapid development and for testing error
+handling code. To support host-based tests, `libtock_core` must be built out of
+portable pieces that can be unit tested, even though the library as a whole is
+not portable.
```

Comment:
I'm a little confused, do you mean the inbuilt Rust unit tests or some other testing mechanism?

Reply:
Nothing I wrote is specific to Rust's builtin test harness, but yes I intend for host-based unit tests to use that mechanism. Is there something about what I wrote that implies I am or am not referring to Rust's builtin test harness?

Reply:
I was just unsure if you meant mocking a full Tock kernel (like https://github.com/tock/tock/pull/2111) or doing unit tests.

Reply:
Oh, I see. In the **Host-based unit tests** bullet above (currently line 47), I mention using a fake Tock kernel instead of a real Tock kernel.

Reply:
Yep, that is why I was confused.

Reply:
Alistair, is there still confusion remaining?